lib/polygeist/Passes/ParallelLoopDistribute.cpp

```diff
@@ -972,46 +1092,246 @@ struct RotateWhile : public OpRewritePattern<scf::WhileOp> {
   }
 };
 
-/// Splits a parallel loop around the first barrier it immediately contains.
-/// Values defined before the barrier are stored in newly allocated buffers and
-/// loaded back when needed.
-template <typename T>
-struct DistributeAroundBarrier : public OpRewritePattern<T> {
-  DistributeAroundBarrier(MLIRContext *ctx) : OpRewritePattern<T>(ctx) {}
-
-  LogicalResult splitSubLoop(T op, PatternRewriter &rewriter, BarrierOp barrier,
-                             SmallVector<Value> &iterCounts, T &preLoop,
-                             T &postLoop, Block *&outerBlock, T &outerLoop,
-                             memref::AllocaScopeOp &outerEx) const;
-
-  LogicalResult matchAndRewrite(T op,
-                                PatternRewriter &rewriter) const override {
-    if (op.getNumResults() != 0) {
-      LLVM_DEBUG(DBGS() << "[distribute] not matching reduction loops\n");
-      return failure();
-    }
-
-    if (!isNormalized(op)) {
-      LLVM_DEBUG(DBGS() << "[distribute] non-normalized loop\n");
-      return failure();
-    }
+LogicalResult splitSubLoop(scf::ParallelOp op, PatternRewriter &rewriter,
+                           BarrierOp barrier, SmallVector<Value> &iterCounts,
```

Comment:
Is there a reason why the template for the for loop type was undone?

Reply:
I could not get the partial template application working after I added the bool UseMinCut template argument. now it is an overloaded function living outside the class